### PR TITLE
feat: webhooks tests in Controller API bddtests

### LINKED ISSUE #545: Build Reference Controller
- Receives webhook events
- Process the events
- Trigger Agent RestAPI (if required)

### LINKED ISSUE #585: Use docker file for docker based agents
Instead of building aries-agentd inside docker command use DockerFile to create pre-built image.

`test/bdd/fixtures/agent/docker-compose.yml`

### PR DESCRIPTION
**Pull request by sudeshrshetty**
- Completed all the verification steps for 'didexchange_e2e_controller'
- Added sample webhook server which maintains queue for pushed topics
- Building agent and sample webhook images before running test
- Fixed minor issues in client found after running tests
- closes #545
- closes #585

Signed-off-by: sudesh.shetty <sudesh.shetty@securekey.com>

### CODE REVIEW

File: pkg/restapi/operation/didexchange/didexchange.go [hunk 1]
```diff
@@ -404,6 +406,7 @@ func (c *Operation) sendConnectionNotification(connectionID string) error {
 		return fmt.Errorf("connection notification json marshal : %w", err)
 	}
 
+	logger.Infof("Sending notification on topic '%s', message body : %s", connectionsWebhookTopic, jsonMessage)
```
Comment (troyronda) [inline pkg/restapi/operation/didexchange/didexchange.go]: Why Info level?

Comment (sudeshrshetty) [reply pkg/restapi/operation/didexchange/didexchange.go]: Changed it to DEBUG

File: test/bdd/bddtests_test.go [hunk 2]
```diff
@@ -108,6 +108,8 @@ func getCmdArg(argName string) string {
 
 func FeatureContext(s *godog.Suite) {
 
+	log.SetLevel("", log.DEBUG)
```
Comment (troyronda) [inline test/bdd/bddtests_test.go]: @fqutishat aren't we passing log levels in?

Comment (sudeshrshetty) [reply test/bdd/bddtests_test.go]: Reverted this line, enabled it to see logs locally.

File: go.mod [hunk 3]
```diff
@@ -8,20 +8,21 @@ require (
 	github.com/DATA-DOG/godog v0.7.13 // indirect
 	github.com/agl/ed25519 v0.0.0-20170116200512-5312a6153412
 	github.com/btcsuite/btcutil v0.0.0-20190425235716-9e5f4b9a998d
+	github.com/golang/protobuf v1.3.0 // indirect
```
Comment (fqutishat) [inline go.mod]: we don't need

File: go.mod [hunk 4]
```diff
@@ -8,20 +8,21 @@ require (
...
-	github.com/sirupsen/logrus v1.4.2 // indirect
 	github.com/spf13/cobra v0.0.5
 	github.com/square/go-jose/v3 v3.0.0-20190722231519-723929d55157
 	github.com/stretchr/testify v1.3.0
 	github.com/syndtr/goleveldb v1.0.0
 	github.com/xeipuuv/gojsonpointer v0.0.0-20180127040702-4e3ac2762d5f // indirect
 	github.com/xeipuuv/gojsonreference v0.0.0-20180127040603-bd5ef7bd5415 // indirect
 	github.com/xeipuuv/gojsonschema v1.1.0
-	golang.org/x/crypto v0.0.0-20190701094942-4def268fd1a4
+	golang.org/x/crypto v0.0.0-20190927123631-a832865fa7ad
```
Comment (fqutishat) [inline go.mod]: we don't need this import

File: images/webhook/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,30 @@
+#
```
Comment (troyronda) [inline images/webhook/Dockerfile]: This is a test/mock image?

Perhaps we should have a parent folder to make that more obvious?

File: test/bdd/bddtests_test.go [hunk 6]
```diff
@@ -61,7 +61,7 @@ func TestMain(m *testing.M) {
 					composition = append(composition, newComposition)
 				}
 				fmt.Println("docker-compose up ... waiting for containers to start ...")
-				testSleep := 15
+				testSleep := 5
```
Comment (troyronda) [inline test/bdd/bddtests_test.go]: is 5s reliably enough time?

File: images/agent/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,30 @@
+#
+# Copyright SecureKey Technologies Inc. All Rights Reserved.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+ARG GO_VER
+ARG ALPINE_VER
+
+FROM golang:${GO_VER}-alpine${ALPINE_VER} as golang
```
Comment (troyronda) [inline images/agent/Dockerfile]: TODO: The image should not require golang. Instead it should be from base alpine.
Build the binary outside and copy into the image.

(the image is now much bigger than it needs to be.)

File: pkg/restapi/operation/didexchange/didexchange.go [hunk 8]
```diff
@@ -242,9 +242,10 @@ func (c *Operation) QueryConnections(rw http.ResponseWriter, req *http.Request)
 //        200: queryConnectionResponse
 func (c *Operation) QueryConnectionByID(rw http.ResponseWriter, req *http.Request) {
 	params := mux.Vars(req)
-	logger.Debugf("Querying connection invitation for id [%s]", params["id"])
+	logger.Infof("Querying connection invitation for id [%s]", params["id"])
```
Comment (troyronda) [inline pkg/restapi/operation/didexchange/didexchange.go]: Looks like log levels need to be cleaned up in this PR.

File: test/bdd/webhook/main.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	select {
+	case topic := <-connectionTopics:
+		w.Write(topic)
+	case <-time.After(100 * time.Millisecond):
+		fmt.Fprintf(w, `{"error":"no topic found in queue"}`)
+	}
+}
+
+func main() {
+	port := os.Getenv("PORT")
```
Comment (troyronda) [inline test/bdd/webhook/main.go]: This should be a more specific env variable name.

File: test/bdd/bddtests_test.go (hunk 6 above)
Comment (sudeshrshetty) [reply test/bdd/bddtests_test.go]: I reverted my previous change where docker-compose command script was building and launching `aries-agentd`, now we are building image beforehand in new make targets I added, so start time of 5s is enough.

File: images/agent/Dockerfile (hunk 7 above)
Comment (sudeshrshetty) [reply images/agent/Dockerfile]: You mean building with `env GOOS=target-OS GOARCH=target-architecture` in host machine. 

Created #606 , will address in upcoming PR

File: test/bdd/webhook/main.go (hunk 9 above)
Comment (sudeshrshetty) [reply test/bdd/webhook/main.go]: done

File: pkg/restapi/operation/didexchange/didexchange.go (hunk 8 above)
Comment (sudeshrshetty) [reply pkg/restapi/operation/didexchange/didexchange.go]: done

File: images/webhook/Dockerfile (hunk 5 above)
Comment (sudeshrshetty) [reply images/webhook/Dockerfile]: done

File: go.mod (hunk 4 above)
Comment (sudeshrshetty) [reply go.mod]: done

File: go.mod (hunk 3 above)
Comment (sudeshrshetty) [reply go.mod]: done

File: images/agent/Dockerfile (hunk 7 above)
Comment (fqutishat) [reply images/agent/Dockerfile]: @sudeshrshetty we should not build the binary in the host machine

Comment (fqutishat) [reply images/agent/Dockerfile]: we just to make sure that generated image not include golang but the build should still inside in image

Comment (fqutishat) [reply images/agent/Dockerfile]: @sudeshrshetty can you run the generated image and check if golang is exists

Comment (fqutishat) [reply images/agent/Dockerfile]: @troyronda @sudeshrshetty I think the golang is not exist in the generated image we will confirm

Comment (sudeshrshetty) [reply images/agent/Dockerfile]: I double checked generated images, they don't have golang in it. @fqutishat @troyronda

**Pull request merged**